onnxscript/ir/passes/common/shape_inference.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Shape inference pass using onnx.shape_inference."""
+
+from __future__ import annotations
+
+__all__ = ["ShapeInferencePass"]
+
+import logging
+
+import onnx
+
+from onnxscript import ir
+
+logger = logging.getLogger(__name__)
+
+
+class ShapeInferencePass(ir.passes.PassBase):
+    """This pass performs shape inference on the graph."""
+
+    # This pass does not modify the model in place.
+    in_place = False
+
+    def __init__(
+        self, check_type: bool = True, strict_mode: bool = True, data_prop: bool = True
+    ) -> None:
+        """Initialize the shape inference pass.
+
+        Args:
+            check_type: If True, check the types of the inputs and outputs.
+            strict_mode: If True, use strict mode for shape inference.
+            data_prop: If True, use data propagation for shape inference.
+        """
+        super().__init__()
+        self.check_type = check_type
+        self.strict_mode = strict_mode
+        self.data_prop = data_prop
+
+    def call(self, model: ir.Model) -> ir.passes.PassResult:
+        # Store the original initializer values so they can be restored
+        initializer_values = tuple(model.graph.initializers.values())
+        tensors = {v.name: v.const_value for v in initializer_values}
+        original_inputs_len = len(model.graph.inputs)
+        initializer_names = {v.name for v in initializer_values}
+
+        # Turn the initializers into inputs and clear the initializers
```

Comment:
Two points to consider: (a) ONNX does allow large-tensors to be stored as attributes in nodes, and (b) What about small tensor-constants (which can be useful for shape-inference to produce better results)?

So, it would be useful if small tensors are NOT removed from the serialized model that is input to shape-inference (for (b))

For attributes: removing them here is tricky, so may need to handle it some other way (as external tensors?)

Reply:
(b) sounds good. I will add a threshold like we do in ir save
(a) is doable too but depending on how shape inference treats them. I propose we handle them when we see a need?

Reply:
Done (b)